pkg/kvstore/etcd.go

```diff
@@ -0,0 +1,372 @@
+	} else if config != nil {
+		c, err = client.New(*config)
+	} else {
+		return nil, fmt.Errorf("empty configuration provided")
+	}
+	if err != nil {
+		return nil, err
+	}
+	log.Info("Waiting for etcd client to be ready...")
+	s, err := concurrency.NewSession(c)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to contact etcd: %s", err)
+	}
+	log.Info("Etcd client ready")
+	ec := &EtcdClient{
+		Client:      c,
+		session:     s,
+		sessionMU:   &sync.RWMutex{},
+		lockPaths:   map[string]*sync.Mutex{},
+		lockPathsMU: &sync.Mutex{},
+	}
+	go func() {
+		for {
+			select {
+			case <-ec.session.Done():
+				newSession, err := concurrency.NewSession(c)
+				if err != nil {
+					log.Errorf("Error while renewing etcd session %s", err)
+					time.Sleep(10 * time.Second)
+				} else {
+					ec.sessionMU.Lock()
+					ec.session = newSession
+					ec.sessionMU.Unlock()
+					log.Debugf("Renewing etcd session")
+				}
+			}
+		}
+	}()
+	return ec, nil
+}
+
+func (e *EtcdClient) LockPath(path string) (KVLocker, error) {
+	e.lockPathsMU.Lock()
+	if e.lockPaths[path] == nil {
+		e.lockPaths[path] = &sync.Mutex{}
+	}
+	e.lockPathsMU.Unlock()
+
+	log.Debugf("Locking path %s", path)
+	// First we lock the local lock for this path
+	e.lockPaths[path].Lock()
+	e.sessionMU.RLock()
+	mu := concurrency.NewMutex(e.session, path)
+	e.sessionMU.RUnlock()
+	// Then we lock the global lock
+	err := mu.Lock(ctx.Background())
+	if err != nil {
+		e.lockPaths[path].Unlock()
+		return nil, fmt.Errorf("Error while locking path %s: %s", path, err)
+	}
+	log.Debugf("Locked path %s", path)
+	return &EtcdLocker{mutex: mu, path: path, localLock: e.lockPaths[path]}, nil
+}
+
+func (e *EtcdLocker) Unlock() error {
+	err := e.mutex.Unlock(ctx.Background())
+	e.localLock.Unlock()
+	if err == nil {
+		log.Debugf("Unlocked path %s", e.path)
+	}
+	return err
+}
+
+func (e *EtcdClient) GetValue(k string) (json.RawMessage, error) {
+	gresp, err := e.Get(ctx.Background(), k)
+	if err != nil {
+		return nil, err
+	}
+	if gresp.Count == 0 {
+		return nil, nil
+	}
+	return json.RawMessage(gresp.Kvs[0].Value), nil
+}
+
+func (e *EtcdClient) SetValue(k string, v interface{}) error {
+	vByte, err := json.Marshal(v)
+	if err != nil {
+		return err
+	}
+	_, err = e.Put(ctx.Background(), k, string(vByte))
+	return err
+}
+
+func (e *EtcdClient) InitializeFreeID(path string, firstID uint32) error {
+	kvLocker, err := e.LockPath(path)
+	if err != nil {
+		return err
+	}
+	defer kvLocker.Unlock()
+
+	log.Debug("Trying to acquire free ID...")
+	k, err := e.GetValue(path)
+	if err != nil {
+		return err
+	}
+	if k != nil {
+		// FreeID already set
+		return nil
+	}
+	log.Info("Trying to put free ID...")
+	err = e.SetValue(path, firstID)
+	if err != nil {
+		return err
+	}
+	log.Info("Free ID for path %s successfully initialized", path)
+
+	return nil
+}
+
+func (e *EtcdClient) GetMaxID(key string, firstID uint32) (uint32, error) {
+	value, err := e.GetValue(key)
```

Comment:
If you're *really* concerned with that case, use a for loop instead, with a attempts counter check.

```go
func (e *EtcdClient) GetMaxID(key string, firstID uint32) (freeID uint32, err error) {
    attempts := 3
    for {
        switch value, err := e.GetValue(key); {
        case attempts == 0:
            err = errors.New("Unable to retreive last free ID because key is always empty")
            log.Error(err)
            fallthrough
        case err != nil:
            return 0, err
        case value == nil:
            if err := e.InitializeFreeID(key, firstID); err != nil {
                return 0, err
            }
            attempts--
        }
    }

    if err := json.Unmarshal(value, &freeID); err != nil {
        return 0, err
    }

    log.Debugf("Retrieving max free ID %d", freeID)
    return freeID, nil
}
```